### PR TITLE
Prototype sampling-based motion planning + 3d representations

### PR DESCRIPTION
**Pull request by cpaxton**
## Motivation and Context

- add lidar to improve real world perf
- add sampling based planners to let us move around through tight clutter
- improve grasping on robot

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: src/home_robot/home_robot/motion/base.py
Comment (ykarmesh) [inline src/home_robot/home_robot/motion/base.py:17]:
```suggestion
    """Stores motion plan. Can be extended."""
```

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:164]: Is this condition ever going to be true? Line 153-154 already do the reshaping and the lines between 156-162 aren't going to change the shapes of those arrays

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:168]: Can we use a class for the observation at timestep t, rather than a tuple with 6 values?

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:235]: Incomplete sentence

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:248]: I think this function still has the error that I was previously mentioning. Maybe it was never noticed, because this function was never used. The problem with this function is that `self.xyz` and `self.feats` are `None` after the call to the `self.reset_cache` call and they will remain `None` throughout this for loop. So at every iteration, the `if statement` will be evaluated to `True` and finally when you reach the `voxel_down_sample` line, pc_xyz and pc_rgb` will contain only the point cloud from last observation.

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:263]:
```suggestion
        # Gets the xyz correctly - for now everything is assumed to be within the correct distance of origin
```

File: src/home_robot/home_robot/mapping/voxel.py
Comment (ykarmesh) [inline src/home_robot/home_robot/mapping/voxel.py:275]: The second line overrides what is written in the first line

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:248]: I should remove this function and just repeatedly call "add"

File: src/home_robot/home_robot/motion/space.py
Comment (cpaxton) [reply src/home_robot/home_robot/motion/space.py:58]: Its ranges, picked for same length as the others. Can change it.

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (yvsriram) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:73]: `InstanceMemory.process_instances_for_env` performs the same thing.

File: src/home_robot/home_robot/mapping/voxel.py
Comment (yvsriram) [inline src/home_robot/home_robot/mapping/voxel.py:118]: We should create a single `InstanceMemory` object instead. This should handle the addition of temporary local instance views (`process_instances_for_env`), and mapping local views to global instances (`update_instance_id`).

File: src/home_robot/home_robot/motion/rrt.py
Comment (cpaxton) [reply src/home_robot/home_robot/motion/rrt.py:30]: thanks tfixed

File: src/home_robot/home_robot/motion/rrt.py
Comment (cpaxton) [reply src/home_robot/home_robot/motion/rrt.py:57]: removed the one in RRT actually, it's wrong

File: src/home_robot/home_robot/motion/space.py
Comment (cpaxton) [reply src/home_robot/home_robot/motion/space.py:58]: update_bounds is called in __init__ so this is fine

Comment (cpaxton) [reply src/home_robot/home_robot/motion/space.py:58]: renamed

File: src/home_robot/home_robot/core/interfaces.py
Comment (alexsax) [inline src/home_robot/home_robot/core/interfaces.py:225]: Should instance also have scores? That's needed for NMS. 
What about category ID?

Maybe "instance image" should be a dataclass bundling these together?

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:235]: fixed

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:168]: done

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:275]: yep, first line is old code (doesnt track floor space)

File: src/home_robot/home_robot/core/interfaces.py
Comment (cpaxton) [reply src/home_robot/home_robot/core/interfaces.py:225]: this is just an instance mask prediction - it already has category segmentations

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:128]: I think the reason i didnt do this is because i dont have a global semantic map?

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:73]: Ah maybe i was misinterpreting the input parameters here?

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:51]: moved this

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:116]: removed it

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:137]: added code to use this; now we need to gfix it

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:213]: removed

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:56]: I thought about it for a bit but it ended up looking really clunky after i moved it *out* of this where it knows what the grid_size actually is, so i think i'll keep it as is

File: src/home_robot/home_robot/mapping/voxel.py
Comment (yvsriram) [inline src/home_robot/home_robot/mapping/voxel.py:230]: Will the `_no_instance_memory` option be removed after this is fixed?

Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:230]: Yeah I think we should do that

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (yvsriram) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:49]: Duplicate? I merged this separately to main

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (yvsriram) [inline src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:73]: Repetition

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py]: torch support. added a comment.

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:49]: removed

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:73]: removed

File: src/home_robot/home_robot/core/interfaces.py
Comment (cpaxton) [reply src/home_robot/home_robot/core/interfaces.py:225]: I think this is actually not a bad idea, maybe we should add as an issue and do it when we need it?

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:74]: Resolving this because i think we have it working now

Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:74]: Can this be removed now?

**Review by yvsriram — APPROVED** [review]
LGTM, once redundant changes are removed from `instance_tracking_modules.py`

File: src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/instance_tracking_modules.py:74]: Seems like this is unused though. May remove this now/later?

File: src/third_party/habitat-lab
Comment (yvsriram) [inline src/third_party/habitat-lab:1]: Please remove this before merging

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:108]: Some of these comments are out of date, feature no longer is RGB and doesnt assume 3 channels.

For factoring out the 2D code - we could do this, not sure what the best way is of handling that

File: src/third_party/habitat-lab
Comment (cpaxton) [reply src/third_party/habitat-lab:1]: done

File: src/home_robot/home_robot/mapping/voxel.py
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:253]: Can we add a param `background_instance_id: int` to process_instances_for_env? This removes a hard-coded/hidden parameter in that function, and saves us from having to work around it here :)

File: src/home_robot/home_robot/mapping/voxel.py
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:199]: Is obs truly optional here if we need it for the instance mask?

File: src/home_robot/home_robot/utils/image.py
Comment (alexsax) [inline src/home_robot/home_robot/utils/image.py:28]: Maybe not in this PR, but should this class be called PinholeCamera since it uses a pinhole model w/o any distortion?

**Review by alexsax — APPROVED** [review]
Minor changes but LGTM

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:234]: done

File: src/home_robot/home_robot/mapping/voxel.py
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:180]: technically you dont need instance masks right?

File: src/home_robot/home_robot/utils/image.py
Comment (cpaxton) [reply src/home_robot/home_robot/utils/image.py:28]: Yeah makes sense - although i was imagining you could override it, so might just note that pinhole is by default

File: src/home_robot/home_robot/utils/image.py
Comment (cpaxton) [reply src/home_robot/home_robot/utils/image.py:24]: done

File: src/home_robot/home_robot/utils/image.py
Comment (cpaxton) [reply src/home_robot/home_robot/utils/image.py:25]: done

**Pull request merged**
